validator/client/wait_for_activation.go

```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
-				return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
-			}
-
-			for _, s := range res.Statuses {
-				v.pubkeyToStatus[bytesutil.ToBytes48(s.PublicKey)] = &validatorStatus{
-					publicKey: s.PublicKey,
-					status:    s.Status,
-					index:     s.Index,
-				}
-			}
-
-			// "-1" indicates that validator count endpoint is not supported by the beacon node.
-			var valCount int64 = -1
-			valCounts, err := v.prysmChainClient.ValidatorCount(ctx, "head", []validator2.Status{validator2.Active})
-			if err != nil && !errors.Is(err, iface.ErrNotSupported) {
-				return errors.Wrap(err, "could not get active validator count")
-			}
-
-			if len(valCounts) > 0 {
-				valCount = int64(valCounts[0].Count)
-			}
-
-			someAreActive = v.checkAndLogValidatorStatus(valCount)
+func (v *validator) waitForAccountsChange(ctx context.Context, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	select {
+	case <-ctx.Done():
+		log.Debug("Context closed, exiting fetching validating keys")
+		return ctx.Err()
+	case <-accountsChangedChan:
+		// If the accounts changed, try again.
+		return v.internalWaitForActivation(ctx, accountsChangedChan)
+	}
+}
+
+func (v *validator) waitForActivationRetry(ctx context.Context, span octrace.Span, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	select {
+	case <-ctx.Done():
+		log.Debug("Context closed, exiting fetching validating keys")
+		return ctx.Err()
+	case <-accountsChangedChan:
+		// Accounts (keys) changed, restart the process.
+		return v.internalWaitForActivation(ctx, accountsChangedChan)
+	default:
+		log.Warn("No active validator keys provided. Waiting until next epoch to check again...")
+		headSlot, err := v.CanonicalHeadSlot(ctx)
+		if err != nil {
+			return v.handleReconnection(ctx, span, err, "Failed to get head slot. Reconnecting...", accountsChangedChan)
 		}
+		if err := v.waitForNextEpoch(ctx, headSlot, v.genesisTime, accountsChangedChan); err != nil {
+			return v.handleReconnection(ctx, span, err, "Failed to wait for next epoch. Reconnecting...", accountsChangedChan)
+		}
+		return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
 	}
+}
 
-	return nil
+// waitForNextEpoch creates a blocking function to wait until the next epoch start given the current slot
+func (v *validator) waitForNextEpoch(ctx context.Context, slot primitives.Slot, genesisTimeSec uint64, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	firstSlotOfNextEpoch, err := slots.EpochStart(slots.ToEpoch(slot) + 1)
+	if err != nil {
+		return err
+	}
+	nextEpochStartDuration, err := slots.ToTime(genesisTimeSec, firstSlotOfNextEpoch)
+	if err != nil {
+		return err
+	}
+	ss, err := slots.SecondsSinceSlotStart(slot, genesisTimeSec, uint64(time.Now().Unix()))
+	if err != nil {
+		return err
+	}
+	waitTime := uint64(nextEpochStartDuration.Unix()-time.Now().Unix()) + ss
```

Comment:
I think this is incorrect, you should not be adding `ss`. You already know the exact point in time when the next epoch will start, so subtracting the current time from that time gives you the correct answer.

Reply:
I tested it, the time isn't always 100% exact, sometimes it's like 1-2 seconds off when the new epoch starts
the +ss is just from the current time in the slot so it guarantees it is within the bounds of the start of the first epoch. sometimes adding a 1 second buffer doesn't work